runtime/inspector/mod.rs

```diff
@@ -0,0 +1,722 @@
+    let detail = v8::inspector::StringView::empty();
+    self
+      .v8_session
+      .borrow_mut()
+      .as_mut()
+      .unwrap()
+      .schedule_pause_on_next_statement(reason, detail);
+  }
+}
+
+impl v8::inspector::ChannelImpl for WebsocketSession {
+  fn base(&self) -> &v8::inspector::ChannelBase {
+    &self.v8_channel
+  }
+
+  fn base_mut(&mut self) -> &mut v8::inspector::ChannelBase {
+    &mut self.v8_channel
+  }
+
+  fn send_response(
+    &mut self,
+    call_id: i32,
+    message: v8::UniquePtr<v8::inspector::StringBuffer>,
+  ) {
+    self.send_message(Some(call_id), message);
+  }
+
+  fn send_notification(
+    &mut self,
+    message: v8::UniquePtr<v8::inspector::StringBuffer>,
+  ) {
+    self.send_message(None, message);
+  }
+
+  fn flush_protocol_notifications(&mut self) {}
+}
+
+impl Future for WebsocketSession {
+  type Output = ();
+  fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+    self.proxy_rx_handler.poll_unpin(cx)
+  }
+}
+
+/// A local inspector session that can be used to send and receive protocol messages directly on
+/// the same thread as an isolate.
+pub struct InMemorySession {
+  v8_channel: v8::inspector::ChannelBase,
+  v8_session: v8::UniqueRef<v8::inspector::V8InspectorSession>,
+  response_tx_map: HashMap<i32, oneshot::Sender<serde_json::Value>>,
+  next_message_id: i32,
+  notification_queue: Vec<Value>,
+}
+
+impl v8::inspector::ChannelImpl for InMemorySession {
+  fn base(&self) -> &v8::inspector::ChannelBase {
+    &self.v8_channel
+  }
+
+  fn base_mut(&mut self) -> &mut v8::inspector::ChannelBase {
+    &mut self.v8_channel
+  }
+
+  fn send_response(
+    &mut self,
+    call_id: i32,
+    message: v8::UniquePtr<v8::inspector::StringBuffer>,
+  ) {
+    let raw_message = message.unwrap().string().to_string();
+    let message: serde_json::Value = match serde_json::from_str(&raw_message) {
+      Ok(v) => v,
+      Err(error) => match error.classify() {
+        serde_json::error::Category::Syntax => json!({
+          "id": call_id,
+          "result": {
+            "result": {
+              "type": "error",
+              "description": "Unterminated string literal",
+              "value": "Unterminated string literal",
+            },
+            "exceptionDetails": {
+              "exceptionId": 0,
+              "text": "Unterminated string literal",
+              "lineNumber": 0,
+              "columnNumber": 0
+            },
+          },
+        }),
+        _ => panic!("Could not parse inspector message"),
+      },
+    };
+
+    self
+      .response_tx_map
+      .remove(&call_id)
+      .unwrap()
+      .send(message)
+      .unwrap();
+  }
+
+  fn send_notification(
+    &mut self,
+    message: v8::UniquePtr<v8::inspector::StringBuffer>,
+  ) {
+    let raw_message = message.unwrap().string().to_string();
+    let message = serde_json::from_str(&raw_message).unwrap();
+
+    self.notification_queue.push(message);
```

Comment:
(I realize that this code already existed before this PR)

Pushing the message to a `Vec` without somehow waking up the tokio task could potentially cause a hang. Somewhere in the process someone is "waiting" for notifications, they're not going to magically find out when one has arrived, something needs to be done to wake up that task so they get polled by the event loop.

Reply:
I'm aware of this problem - I've got a WIP commit that has a "pump" future for `InMemorySession` that does that - unfortunately it currently is racy and deadlocks; needs more investigation.